File: leetcode/editor/cn/20_有效的括号_test.go
package leetcode

import (
	"container/list"
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when20_thenSuccess(t *testing.T) {
	assert.Equal(t, true, isValid("()"))
	assert.Equal(t, false, isValid("(]"))
	assert.Equal(t, true, isValid("([])"))
}

//leetcode submit region begin(Prohibit modification and deletion)
func isValid(s string) bool {
	ans := true
	pairs := map[rune]rune{
		'(': ')',
		'[': ']',
		'{': '}',
	}
	stack := list.New()
	for _, r := range s {
		// 判断 key 是否存在
		if _, exists := pairs[r]; exists {
			stack.PushBack(r)
		} else {
			// go nil
			// back.Value 为 any (interface{})类型，强制转换为 rune 类型
			if stack.Len() == 0 || r != pairs[stack.Remove(stack.Back()).(rune)] {
				ans = false
				break
			}
		}
	}
	if stack.Len() != 0 {
		ans = false
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/20_有效的括号_test.py
import unittest


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        self.assertTrue(self.solution.isValid("()"))
        self.assertFalse(self.solution.isValid("[[]"))
        self.assertFalse(self.solution.isValid("[]]"))
        self.assertFalse(self.solution.isValid("]]]"))


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def isValid(self, s: str) -> bool:
        ans: bool = True
        pairs: dict[str, str] = {
            "(": ")",
            "[": ']',
            "{": "}"
        }
        stack: list[str] = []
        for c in s:
            # watch out: can not use pairs[c]
            if pairs.get(c) is not None:
                stack.append(c)
            elif len(stack) == 0 or c != pairs[stack.pop()]:
                ans = False
                break
        if len(stack) != 0:
            ans = False
        return ans


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
